library(patchwork)
library(gridExtra)
library(RColorBrewer)
library(circlize) 
library(scatterpie)
library(tanggle)
library(RSplitsTree)
# library(ggmap)
# library(lubridate)
library(ggplot2)
library(dplyr)
library(data.table)
library(ggrepel)
library(tidyverse)
library(ggsn)
library(ggspatial)
library(readxl)
library(heatmaply)
library(circlize)
library(ComplexHeatmap)
library(igraph)
library(SNPRelate)
library(geosphere)
library(reshape2)
library(vegan)
# library(ggforce)
library(openxlsx)
# library(stringr)
# library(pracma)
library(ggpubr)
library(RRtools)
library(ggthemes)
library(RColorBrewer)
library(ozmaps)
library(adegenet)
library(tidyr)
library(diveRsity)
library(RRtools)
library(LEA)

theme_set(theme_few())



setup_variables <- read.xlsx("0_setup_variables.xlsx", colNames = TRUE)
maindir <- setup_variables[1, 2]
species <- setup_variables[2, 2]
dataset <- setup_variables[3, 2]
RandRbase <- ""
raw_meta_path <- setup_variables[4, 2]
species_col_name <- setup_variables[5, 2]
site_col_name <- setup_variables[6, 2] # this is the equivalent of analysis
remove_pops_less_than_n5 <- setup_variables[7, 2]
downsample <- setup_variables[8, 2]
samples_per_pop <- setup_variables[9, 2] %>% as.numeric()
locus_miss <- setup_variables[10, 2] %>% as.numeric()
sample_miss <- setup_variables[11, 2] %>% as.numeric()
maf_val <- setup_variables[12, 2] %>% as.numeric()
clonal_threshold <- setup_variables[13, 2] %>% as.numeric()
custom_meta <- setup_variables[14, 2]

setwd(maindir)

topskip   <- 6
nmetavar  <- 18

subsubdirectories <- c(paste0("outputs_",site_col_name,"_",species_col_name,"/plots"),
                       paste0("outputs_",site_col_name,"_",species_col_name,"/tables"),
                       paste0("outputs_",site_col_name,"_",species_col_name,"/r_files"))

# Check and create subdirectories if they don't exist
for (subdir in subsubdirectories) {
  
  subdirectory_path <- file.path(paste0(species,'/',subdir))
  
  if (!dir.exists(subdirectory_path)) {
    dir.create(subdirectory_path, recursive = TRUE)
    cat(paste("Created directory:", subdirectory_path, "\n"))
  } else {
    cat(paste("Directory already exists:", subdirectory_path, "\n"))
  }
}
devtools::source_url("https://github.com/eilishmcmaster/SoS_functions/blob/main/sos_functions.R?raw=TRUE")

###note that this loads scripts generated by SY to automate / make RnR summary reports from DArTseq data. Should probably be put onto Git#####
# load("PSFsaved_scripts4autom2.RData")

#### Import and QC ####

#Load data and quality filter loci and samples
d1        <- new.read.dart.xls.onerow(RandRbase,species,dataset,topskip, euchits=FALSE, altcount=TRUE)
qc1       <- report.dart.qc.stats(d1, RandRbase, species, dataset, threshold_missing_loci = 0.8)

d2        <- remove.poor.quality.snps(d1, min_repro=0.96, max_missing=locus_miss)
qc2       <- report.dart.qc.stats(d2, RandRbase, species, dataset)

d3        <- sample.one.snp.per.locus.random(d2, seed=214241)
qc3       <- report.dart.qc.stats(d3, RandRbase, species, dataset)

#Load meta data and attach to dart data
m1        <- read.meta.data.full.analyses.df(d3, RandRbase, species, dataset)

write.table(data.frame(sample=d3$sample_names[!(d3$sample_names %in% m1$sample_names)]), 
            paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/tables/samples_in_dart_not_in_meta.tsv"), sep="\t", row.names = FALSE)

dm        <- dart.meta.data.merge(d3, m1)

# remove samples that are NA for site variable 
samples_with_site_variable <- dm$sample_names[!is.na(dm$meta$analyses[,site_col_name])]
dms2 <- remove.by.list(dm, samples_with_site_variable)



dms <- dms2
treatment <- dms$treatment
# get mean values of lat and long per species site 

# # Original Site Summary
unfiltered_site_summary <- dms2$meta$analyses %>% as.data.frame()%>%
  group_by(!!sym(species_col_name), !!sym(site_col_name)) %>%
  summarize(n_unfiltered = sum(n()),
            lat = mean(as.numeric(lat), na.rm=TRUE),
            long = mean(as.numeric(long),na.rm=TRUE),
            .groups = 'drop') %>%
  filter(n_unfiltered > 0) %>%
  as.data.frame()


####################################### Remove populations with less than five samples ####################################### 

samples_high_missing <- dms$sample_names[which(rowMeans(is.na(dms$gt))>sample_miss)]

write.table(data.frame(sample=samples_high_missing, sample_miss=rowMeans(is.na(dms$gt))[which(rowMeans(is.na(dms$gt))>sample_miss)]), 
            paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/tables/high_missing_samples_removed.tsv"), sep="\t", row.names = FALSE)


dms <- remove.by.missingness(dms, sample_miss)
sites_high_missing_samples_removed <- table(dms$meta$analyses[,site_col_name])

if(remove_pops_less_than_n5=="TRUE"){
  not_n5_sites <- as.vector(names(which(sites_high_missing_samples_removed<samples_per_pop))) #remove groups where n<=1
  not_n5_samples <- dms$sample_names[which(!(dms$meta$analyses[,site_col_name] %in% not_n5_sites))]
  dms <- remove.by.list(dms, not_n5_samples)
  
  dms_no_n1_sites <- dms
}else{
  # make dms where there are no n=1 site
  not_n1_sites <- as.vector(unfiltered_site_summary[unfiltered_site_summary$n_unfiltered<=1,2]) #remove groups where n<=1
  not_n1_samples <- dms$sample_names[which(!(dms$meta$analyses[,site_col_name] %in% not_n1_sites))]
  dms_no_n1_sites <- remove.by.list(dms, not_n1_samples)
}

# #####Subsamples populations down to a standard number of samples (change value in third line to change target value)######
if(downsample=="TRUE"){
  n5_samples_df <- dms$meta$analyses %>% as.data.frame() %>%
    group_by(!!sym(site_col_name)) %>% slice_sample(n = samples_per_pop)
  dms <- remove.by.list(dms,n5_samples_df$sample)
}

####################################### Summarise samples by site ####################################### 

filtered_site_summary <- dms$meta$analyses %>% as.data.frame()%>%
  group_by(!!sym(species_col_name), !!sym(site_col_name)) %>%
  summarize(n_filtered = sum(n()),
            lat = mean(as.numeric(lat), na.rm=TRUE),
            long = mean(as.numeric(long),na.rm=TRUE),
            .groups = 'drop') %>%
  as.data.frame()


site_samples_summary <- merge(unfiltered_site_summary, filtered_site_summary[,1:3],
                              by = c(site_col_name, species_col_name), all.x = TRUE)


site_samples_summary <- site_samples_summary[,c(2,1,4,5,3,6)]
site_samples_summary$n_filtered[is.na(site_samples_summary$n_filtered)] <- 0

site_categories <- unique(dms2$meta$analyses[,site_col_name])
site_categories <- site_categories[site_categories!="no_geo_data"]
site_categories <- c(site_categories[order(as.numeric(site_categories))], "no_geo_data")

site_samples_summary <-site_samples_summary[order(match(site_samples_summary[,site_col_name], site_categories)),]

# Calculate total values per unique site_samples_summary[,1] value
total_summary <- site_samples_summary %>%
  group_by(!!sym(species_col_name)) %>%
  summarize(n_unfiltered = sum(n_unfiltered, na.rm=TRUE), n_filtered = sum(n_filtered, na.rm=TRUE))

total_summary[site_col_name] <- "Total"

# Combine the original and total summaries
final_summary <- bind_rows(site_samples_summary, total_summary)

write.xlsx(final_summary, paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/tables/SUMMARY_",site_col_name,"_sample_numbers.xlsx"), asTable = FALSE, overwrite = TRUE)


###########################################  Colour palettes ###########################################  


sp_colours <- named_list_maker(dms$meta$analyses[,species_col_name], "Spectral", 9)

sp_shapes <- 1:length(unique(dms$meta$analyses[,species_col_name]))
names(sp_shapes) <- unique(dms$meta$analyses[,species_col_name])

if(custom_meta=="FALSE"){
  # get sites in order
  site_categories2 <- unique(dms$meta$analyses[,site_col_name])
  site_categories2 <- site_categories2[site_categories2!="no_geo_data"]
  site_categories2 <- c(site_categories2[order(as.numeric(site_categories2))], "no_geo_data")
  
  site_colours <- named_list_maker(site_categories2, "Set3", 12)
  site_colours['no_geo_data'] <- "grey30"
  
  site_shapes <- rep(1:25, length.out = length(site_categories2))
  names(site_shapes) <- unique(site_categories2)
  
  site_order <- names(site_colours)
}else{
  # get sites in order
  site_categories2 <- unique(dms$meta$analyses[,site_col_name])
  site_colours <- named_list_maker(site_categories2, "Set3", 12)
  site_shapes <- rep(1:25, length.out = length(site_categories2))
  names(site_shapes) <- unique(site_categories2)
  
  site_order <- names(site_colours)
}

site_labels <-   geom_text_repel(data=filtered_site_summary,
                                 mapping=aes(x=long, y=lat, label=!!sym(site_col_name)),colour="black",
                                 size=3, max.overlaps=10, show.legend=FALSE,force_pull = 2, box.padding = 0.25, segment.size=0.1, min.segment.length = 0)

####################################### MAP ######################################

divxlims <- c(min(filtered_site_summary$long, na.rm=TRUE)-0.1,
              max(filtered_site_summary$long, na.rm=TRUE)+0.1) #find the min / max longitude
divylims <- c(min(filtered_site_summary$lat, na.rm=TRUE)-0.1,
              max(filtered_site_summary$lat, na.rm=TRUE)+0.1) #find the min / max latitude

base_map <- ggplot(ozmaps::abs_ste) + geom_sf(fill="grey96", colour="grey28") +
  coord_sf(xlim = divxlims, ylim = divylims) + labs(y=element_blank(), x=element_blank())+
  theme(axis.text.x = element_text(angle=90, size=6),axis.text.y=element_text(size=6))+
  ggsn::scalebar(filtered_site_summary[!is.na(filtered_site_summary$lat),], dist = round(diff(divxlims)*20,-1), dist_unit = "km", location = "bottomright", 
                 st.bottom = F, st.size = 2, st.dist = 0.02,border.size =0.5,
                 transform = TRUE, model = "WGS84", height = 0.01)

site_map <- base_map+
  geom_point(data=filtered_site_summary,
             mapping=aes(x=long, y=lat, fill=!!sym(site_col_name)), shape=21, size=2)+ 
  theme(legend.key.size = unit(0, 'lines'), legend.position="bottom")+
  scale_fill_manual(values=site_colours) + site_labels +theme(legend.box = "vertical")+
  if (length(site_categories2) > 20) {
    theme(legend.position = "none")
  }

species_map <- base_map+
  geom_point(data=filtered_site_summary,
             mapping=aes(x=long, y=lat, fill=!!sym(species_col_name)), shape=21, size=2)+ 
  theme(legend.key.size = unit(0, 'lines'), legend.position="bottom", legend.direction = "vertical")+
  scale_fill_manual(values=sp_colours) + site_labels+
  guides(fill = guide_legend(ncol = 2))

###########################################  clones ########################################### 
# find and remove clones using SNPrelate kinship 

#calculate kinship by population 
# VERY important that the population groups are true genetic groups and not conglomerates of multiple genetic groups
kin <- individual_kinship_by_pop(dms, RandRbase, species, dataset, dms$meta$analyses[,species_col_name], maf=0.1, mis=locus_miss, as_bigmat=TRUE)
kin[is.na(kin)] <- 0
# Finding the clones
kin3 <- ifelse(kin < clonal_threshold, 0, 1)

# cluster the clones 
network <- graph_from_adjacency_matrix(kin3, mode="undirected", diag=F,weighted=T) #makes the network based on k>0.45 

ceb <- cluster_fast_greedy(network) # makes cluster groupings

# get clones 
clones <-as.data.frame(cbind(genet=ceb$membership, sample=ceb$names)) #get the clones from the network as a df
clones_out <- merge(clones, dms$meta$analyses[,c("sample","lat","long",site_col_name,species_col_name)], by="sample") #add some metadata
clones_out <- clones_out[order(as.numeric(clones_out$genet)),] #order the table by genet 
clones_out$genet <- as.numeric(clones_out$genet)

write.xlsx(clones_out, paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/tables/PLINK_clones.xlsx"), asTable = FALSE, overwrite = TRUE)

###########################################  kinship heatmap ####################################### 
#### Kinship by distance ####

col_fun1 = colorRamp2(c(0,0.2,0.4), c("white", "red","black"))


# kinship_df <- dist_kinship_matrix(dms$gt) %>% as.data.frame(.)
# kinship_df$sample <- rownames(kinship_df)

kinship_df <- kin %>% as.data.frame()
kinship_df$sample <- rownames(kinship_df)


kin_heatmap <- merge(kinship_df, dms$meta$analyses[,c("sample", "lat", species_col_name, site_col_name)],
                   by="sample", all.x=TRUE, all.y=FALSE)
kin_heatmap <- kin_heatmap[match(rownames(kinship_df),kin_heatmap$sample),]
rownames(kin_heatmap) <- kin_heatmap[,"sample"]

kin_heatmap[,"sample"] <- NULL

kin_heatmap[kin_heatmap[,site_col_name]=="no_geo_data",site_col_name] <- NA
kin_heatmap[,site_col_name] <- as.numeric(kin_heatmap[,site_col_name])

kin_heatmap <- kin_heatmap[order(kin_heatmap[,site_col_name]),
                       c(order(kin_heatmap[,site_col_name]),(nrow(kin_heatmap)+1):ncol(kin_heatmap))] #order the table by genet 

#create annotations
site_ann <- HeatmapAnnotation(Site = kin_heatmap[,site_col_name],
                              col = list(Site = site_colours),
                              annotation_name_gp = gpar(fontsize = 0),
                              annotation_legend_param = list(labels_gp=gpar(fontsize=6),#fontface="italic",
                                                             title_gp=gpar(fontsize=8),
                                                             title=site_col_name))


sp_ann <- HeatmapAnnotation(Species = kin_heatmap[,species_col_name],
                            col=list(Species=sp_colours),
                            na_col="white",
                            annotation_name_gp = gpar(fontsize = 0),
                            annotation_legend_param = list(labels_gp=gpar(fontsize=6),#fontface="italic",
                                                           title_gp=gpar(fontsize=8),
                                                           title=species_col_name))


hma <- Heatmap( as.matrix(kin_heatmap[ , c(1:(nrow(kin_heatmap)))]), 
                col=col_fun1, 
                bottom_annotation=c(site_ann, sp_ann),
                name = "PLINK kinship", #title of legend
                row_names_gp = gpar(fontsize = 6),
                column_names_gp = gpar(fontsize = 6),
                row_names_max_width = unit(15, "cm"),
                border_gp = gpar(col = "black", lty = 1),
                column_order=order(as.numeric(kin_heatmap[,site_col_name])),
                row_order=order(as.numeric(kin_heatmap[,site_col_name]))
)


# Set the file name and parameters
heatmap_width <- nrow(kin_heatmap)*0.1 + 7
heatmap_height <- nrow(kin_heatmap)*0.1 + 1

filename <- paste0(species, "/outputs_",site_col_name,"_",species_col_name,"/plots/PLINK_kin_heatmap.pdf")
pdf(filename, width = heatmap_width, height = heatmap_height)
draw(hma, merge_legend = TRUE)
dev.off()


####################################### EUCLIDEAN DISTANCE ####################################### 


col_fun2 = colorRamp2(c(0,0.5,1), c("white", "red","black"))

dist_raw <- as.matrix(dist(dms$gt, diag=TRUE))
distship_df <- 1- (dist_raw/max(dist_raw, na.rm=TRUE)) %>% as.data.frame()
distship_df$sample <- rownames(distship_df)

dist_heatmap <- merge(distship_df, dms$meta$analyses[,c("sample","lat", species_col_name, site_col_name)],
                     by="sample", all.x=TRUE, all.y=FALSE)
dist_heatmap <- dist_heatmap[match(rownames(distship_df),dist_heatmap$sample),]
rownames(dist_heatmap) <- dist_heatmap[,"sample"]

dist_heatmap[,"sample"] <- NULL

dist_heatmap[dist_heatmap[,site_col_name]=="no_geo_data",site_col_name] <- NA
dist_heatmap[,site_col_name] <- as.numeric(dist_heatmap[,site_col_name])

dist_heatmap <- dist_heatmap[order(dist_heatmap[,site_col_name]),
                           c(order(dist_heatmap[,site_col_name]),(nrow(dist_heatmap)+1):ncol(dist_heatmap))] #order the table by genet 

dist_heatmap_plot <- Heatmap( as.matrix(dist_heatmap[ , c(1:(nrow(dist_heatmap)))]), 
                col=col_fun2, 
                bottom_annotation=c(site_ann, sp_ann),
                name = "Inverse\nEuclidean\ndistance", #title of legend
                row_names_gp = gpar(fontsize = 6),
                column_names_gp = gpar(fontsize = 6),
                row_names_max_width = unit(15, "cm"),
                border_gp = gpar(col = "black", lty = 1)
)


# draw(dist_heatmap_plot, merge_legend = TRUE)

# Set the file name and parameters
filename <- paste0(species, "/outputs_",site_col_name,"_",species_col_name,"/plots/EUCLIDEAN_dist_heatmap.pdf")
pdf(filename, width = heatmap_width, height = heatmap_height)
draw(dist_heatmap_plot, merge_legend = TRUE)
dev.off()

################################################ PCA ################################################ 

dms_1000 <- remove.loci.randomly(dms, 1000)
gen_d5 <- new("genlight", dms_1000[["gt"]]) #convert df to genlight object for glPca function
gen_pca <- glPca(gen_d5, parallel=TRUE, nf=5) #do pca -- this method allows the input to have NAs 
g_pca_df <- gen_pca[["scores"]] #extract PCs 
g_pca_df2 <- merge(g_pca_df, dms$meta$analyses, by.x=0, by.y="sample", all.y=FALSE, all.x=FALSE) # add metadata 

pcnames <- paste0(colnames(g_pca_df)," (",
                  paste(round(gen_pca[["eig"]][1:5]/sum(gen_pca[["eig"]]) *100, 2)),
                  "%)") #create names for axes


# species PCA
pca_plot_pc12_species <- ggplot(g_pca_df2, aes(x=PC1, y=PC2, colour=!!sym(species_col_name)))+ 
  geom_point()+xlab(pcnames[1])+ylab(pcnames[2])+
  theme(legend.key.size = unit(0, 'lines'),# legend.position = "right",
        legend.text=element_text(face="italic"))+
  scale_colour_manual(values=sp_colours) 

ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/PCA_",species_col_name,"_PC12.png"), plot = pca_plot_pc12_species, width = 20, height = 15, dpi = 600, units = "cm")

# site PCAs'
pca_labels <- geom_text_repel(colour="black",size=2, max.overlaps=10,
                              show.legend=FALSE,force_pull = 2, box.padding = 0.25, segment.size=0.1, min.segment.length = 0)


pca_plot_pc12_site <- ggplot(g_pca_df2, 
                             aes(x=PC1, y=PC2,
                                 colour=!!sym(site_col_name), label=!!sym(site_col_name)))+ 
  xlab(pcnames[1])+ylab(pcnames[2])+geom_point()+
  pca_labels+  scale_colour_manual(values=site_colours)+ 
  theme(legend.key.size = unit(0, 'lines'))+
  guides(color = guide_legend(nrow = 4))+
  if (length(site_categories2) > 20) {
    theme(legend.position = "none")
  }

pca_plot_pc23_site <- ggplot(g_pca_df2, 
                             aes(x=PC3, y=PC2,
                                 colour=!!sym(site_col_name), label=!!sym(site_col_name)))+ 
  xlab(pcnames[3])+ylab(pcnames[2])+geom_point()+scale_colour_manual(values=site_colours)+ 
  theme(legend.key.size = unit(0, 'lines'))+
  pca_labels

pca_plot_pc34_site <- ggplot(g_pca_df2, 
                             aes(x=PC3, y=PC4,
                                 colour=!!sym(site_col_name), label=!!sym(site_col_name)))+ 
  xlab(pcnames[3])+ylab(pcnames[4])+geom_point()+scale_colour_manual(values=site_colours)+ 
  theme(legend.key.size = unit(0, 'lines'))+
  pca_labels

legend_position <- if (length(site_categories2) > 20) "none" else "bottom"

combined_site_pca <- ggarrange(
  pca_plot_pc12_site, pca_plot_pc23_site, pca_plot_pc34_site,
  ncol = 3, common.legend = TRUE, legend = legend_position)

ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/PCA_",site_col_name,"_all.png"), plot = combined_site_pca, width = 30, height = 10, dpi = 600, units = "cm")
ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/PCA_",site_col_name,"_PC12.png"), plot = pca_plot_pc12_site, width = 20, height = 15, dpi = 600, units = "cm")

# latitude PCAs

pca_plot_pc12_lat <- ggplot(g_pca_df2, aes(x=PC1, y=PC2, colour=as.numeric(lat)))+ 
  labs(color="Latitude")+
  geom_point()+ xlab(pcnames[1])+ylab(pcnames[2])+
  scale_color_gradient(high = "red",low = "blue", na.value = "grey30")

pca_plot_pc23_lat <- ggplot(g_pca_df2, aes(x=PC3, y=PC2, colour=as.numeric(lat)))+ 
  labs(color="Latitude")+
  geom_point()+xlab(pcnames[3])+ylab(pcnames[2])+
  scale_color_gradient(high = "red",low = "blue", na.value = "grey30")

pca_plot_pc34_lat <- ggplot(g_pca_df2, aes(x=PC3, y=PC4, colour=as.numeric(lat)))+ 
  labs(color="Latitude")+
  geom_point()+xlab(pcnames[3])+ylab(pcnames[4])+
  scale_color_gradient(high = "red",low = "blue", na.value = "grey30")

combined_latitude_pca <- ggarrange(pca_plot_pc12_lat, pca_plot_pc23_lat, pca_plot_pc34_lat, ncol=3, common.legend = TRUE, legend="bottom")

ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/PCA_latitude_all.png"), plot = combined_latitude_pca, width = 30, height = 10, dpi = 600, units = "cm")


###########################################  FST ########################################### 

# calculate FST and geodist
gds_file <- dart2gds(dms_no_n1_sites, RandRbase, species, dataset)
pFst      <- population.pw.Fst(dms_no_n1_sites, dms_no_n1_sites$meta$analyses[,site_col_name], RandRbase,species,dataset, maf_val=maf_val, miss_val=locus_miss) #calculates genetic distance 
pS        <- population.pw.spatial.dist(dms_no_n1_sites, dms_no_n1_sites$meta$analyses[,site_col_name]) #calculates geographic distance between populations


####plot IBD plot

# Make self comparisons NA
diag(pFst$Fst) <- NA
diag(pS$S) <- NA

#Mantel test 
man <- mantel(xdis = pS$S, ydis = pFst$Fst, permutations = 10000, na.rm = TRUE) #mantel test, finds if matrices are signficantly similar
man

# mantel plot
Fst_sig <- cbind(melt(pS$S), unlist(as.list(pFst$Fst)))
colnames(Fst_sig)[3] <- "Geo_dist"
colnames(Fst_sig)[4] <- "Fst"
Fst_sig$Geo_dist2 <-Fst_sig$Geo_dist/1000 

# adding metadata for sites
Fst_sig2 <- merge(Fst_sig, distinct(filtered_site_summary[,c(site_col_name,species_col_name)]), by.x="Var1", by.y=site_col_name, all.y=FALSE)
Fst_sig2 <- merge(Fst_sig2, distinct(filtered_site_summary[,c(site_col_name,species_col_name)]), by.x="Var2", by.y=site_col_name, all.y=FALSE)
Fst_sig2$same_sp <- ifelse(Fst_sig2[,6]== Fst_sig2[,7], paste("Within", species_col_name), paste("Between", species_col_name))

fst_manning <- ggplot(Fst_sig2, aes(x= Geo_dist2, y=Fst, color=same_sp))+geom_point(size=1, alpha=0.3)+
  labs(x="Distance (km)", y="FST", colour="Comparison")+
  # facet_zoom(x=Geo_dist2<2, zoom.size=1)+
  geom_hline(yintercept = 0.3, color="black", linetype="dotted")+
  
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), legend.position="bottom")+
  labs(title = paste("Mantel statistic r is", round(man$statistic, 3), ", P =", round(man$signif, 5)))+ theme(plot.title = element_text(size=10))

fst_manning

ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/FST_manning.png"),
       fst_manning, width = 15, height = 10, units = "cm", dpi=600)


# Make heatmaps
# geo dist
geo_d <-pS$S #this is a square matrix
mat <- geo_d/1000 # convert to km 

#FST
mat2 <-pFst$Fst
mat2 <- merge(mat2, filtered_site_summary[,1:2], by.x=0, by.y=site_col_name, all.y=FALSE) #add aggregated df to mat2 (fst)
rownames(mat2) <- mat2$Row.names

mat2$Row.names <- NULL
mat2 <- mat2[match(colnames(mat2)[1:nrow(mat2)],rownames(mat2)),]

order_hm <- Heatmap(mat,
                    cluster_rows = TRUE,
                    cluster_columns = TRUE)
od <- colnames(mat)[column_order(order_hm)]

mat = mat[od, od]
mat2 = mat2[od, c(od,species_col_name)]

# specify fst heatmap colours 
gene_col <-  colorRamp2(c(0,0.5,1), c("#8DD3C7", "white", "#FB8072"))


#specify geo heatmap colours
palette <-  colorRamp2(c(0, max(mat, na.rm=TRUE)), c("white", "#80B1D3"))


row_sp_ann <- rowAnnotation(Species = mat2[,species_col_name],
                                       col=list(Species=sp_colours),
                                       na_col="white",
                                       annotation_legend_param = list(labels_gp=gpar(fontsize=6),#fontface="italic",
                                                                      title_gp=gpar(fontsize=8),
                                                                      title=species_col_name),
                                       annotation_name_gp = gpar(fontsize = 0),
                                       annotation_name_side="top")

bottom_sp_ann <- HeatmapAnnotation(Species = mat2[,species_col_name], 
                                   col = list(Species = sp_colours),
                                              annotation_name_gp = gpar(fontsize = 0),
                                              show_legend = FALSE,
                                              annotation_name_side="right",
                                              na_col = "white",
                                   annotation_legend_param = list(labels_gp=gpar(fontsize=6),#fontface="italic",
                                                                  title_gp=gpar(fontsize=8),
                                                                  title=species_col_name))

geo <- Heatmap(mat,rect_gp = gpar(type = "none"),
               width = nrow(mat)*unit(4, "mm"),
               height = nrow(mat)*unit(4, "mm"),
               col=palette,na_col="white",
               bottom_annotation = bottom_sp_ann,
               column_names_gp = gpar(fontsize = 6),
               cluster_rows = FALSE,
               cluster_columns = FALSE,
               name="Distance (km)",
               heatmap_legend_param = list(title_gp = gpar(fontsize = 8),
                                           labels_gp = gpar(fontsize = 6)),
               cell_fun = function(j, i, x, y, w, h, fill) {
                 if(i >= j) {
                   grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
                 }
               }
)

# make fst heatmap
gene <- Heatmap(as.matrix(mat2[,1:nrow(mat2)]), rect_gp = gpar(type = "none"),
                # column_title = paste(species, "FST and geographic distance"),column_title_gp = gpar(fontsize = 20),#, fontface = "bold"
                width = nrow(mat2)*unit(4, "mm"),
                height = nrow(mat2)*unit(4, "mm"),
                right_annotation = row_sp_ann,
                col=gene_col,na_col="grey",
                row_names_gp = gpar(fontsize = 6),
                column_names_gp = gpar(fontsize = 0),
                border_gp = gpar(col = "black", lty = 1),
                name="FST",
                cluster_rows = FALSE,
                cluster_columns = FALSE,
                heatmap_legend_param = list(title_gp = gpar(fontsize = 8),
                                            labels_gp = gpar(fontsize = 6)),
                cell_fun = function(j, i, x, y, w, h, fill) {
                  if(i <= j) {
                    grid.rect(x, y, w, h, gp = gpar(fill = fill, col = fill))
                    grid.text(sprintf("%.2f", mat2[,1:nrow(mat2)][i, j]), x, y, gp = gpar(fontsize = 4))
                  }
                })

# Set the file name and parameters
filename <- paste0(species, "/outputs_",site_col_name,"_",species_col_name,"/plots/FST_heatmap.pdf")
gene_width <- nrow(mat2)*unit(4, "mm")
pdf(filename, width = (((nrow(mat2)*4)/10) +8)*0.394, height = (((nrow(mat2)*4)/10)+5)*0.394)
draw(geo + gene, ht_gap = -gene_width)
dev.off()

write.xlsx(list(geo_dist_km=mat, fst=mat2), 
            paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/tables/FST_GEODIST_by_",site_col_name,".xlsx"),rowNames = TRUE)


###########################################  Visualise splitstree ########################################### 
## site splitstree
splitstree(dist(dms$gt), paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/r_files/",species,"_",dataset,".nex"))
# 
# Nnet <- phangorn::read.nexus.networx(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/r_files/",species,"_",dataset,".nex"))
# 
# x <- data.frame(x=Nnet$.plot$vertices[,1], y=Nnet$.plot$vertices[,2],
#                 sample=rep(NA, nrow(Nnet$.plot$vertices)))
# 
# x[Nnet$translate$node,"sample"] <- Nnet$translate$label
# x <- merge(x, dms$meta$analyses, by="sample", all.x=TRUE, all.y=FALSE)
# 
# net_x_axis <- max(x$x)-min(x$x)
# net_y_axis <- max(x$y)-min(x$y)
# 
# Nnet$translate$label <-  x[match(Nnet$tip.label, x$sample), site_col_name] %>% .[1:length(Nnet$tip.label)]
# 
# 
# 
# splitstree_plot_site <- ggplot(Nnet, mapping = aes_(~x, ~y), layout = "slanted", mrsd = NULL,
#                                   as.Date = FALSE, yscale = "none", yscale_mapping = NULL,
#                                   ladderize = FALSE, right = FALSE, branch.length = "branch.length",
#                                   ndigits = NULL)+
#   geom_splitnet(layout = "slanted", size=0.2)+
#   geom_point(data=x, aes(x, y, colour=!!sym(site_col_name)))+
#   scale_colour_manual(values=site_colours, na.translate=FALSE,
#                       guide = guide_legend(site_col_name))+
#   geom_tiplab2(size=2, hjust=-2)+
#   theme_void()+
#   expand_limits(x=c(min(x$x)-0.01*net_x_axis, max(x$x)+0.01*net_x_axis),
#                 y=c(min(x$y)-0.01*net_y_axis, max(x$y)+0.01*net_y_axis))+
#   theme(legend.text = element_text(face="italic"), legend.position = "none")+coord_fixed()
# 
# splitstree_plot_site
# 
# 
# ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/small_splitstree.png"), plot = splitstree_plot_site, width = 20, height = 30, dpi = 600, units = "cm")


########################################### LEA ########################################
kvalrange <- 2:7

##check for LEA FOLDER
lea_project <- paste0(maindir,'/',species,"/popgen/",treatment,"/lea/",species,"_",dataset,".snmfProject")

if(!file.exists(lea_project)){
  nd_lea <- dart2lea(dms_no_n1_sites, RandRbase, species, dataset)
  snmf1=snmf(nd_lea, K=kvalrange, entropy = TRUE, repetitions = 1, project = "new")
  plot(snmf1, lwd = 6, col = "red", pch=1)
  # LEA_plots(dms_no_n1_sites, outputloc, nd_lea,treatment, dataset, species)
}else{
  print("LEA project already exists")
}
  
#load whole project
snmf_project <- load.snmfProject(lea_project)

## entropy 
entropy <- t(summary(snmf_project)$crossEntropy) %>% as.data.frame()
entropy_plot <- ggplot(entropy, aes(x=kvalrange, y=mean))+geom_point(colour="red")+
  labs(x="K", y="Mean ncross entropy")+
  theme(legend.position="none")

ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/LEA_entropy.png"), plot = entropy_plot, width = 8, height = 5, dpi = 600, units = "cm")


scatterpie_plots <- list()
admix_bar_plots <- list()
admix_bar_plots_mini <- list()
map_mini <- list()

#for each k value
for (kval in kvalrange){
  ce           <- cross.entropy(snmf_project, K = kval)
  Rbest        <- which.min(ce) # with the lowest cross entropy criterion
  
  qdf <- Q(snmf_project, K=kval, run=Rbest) # get admixture coefficients (Q) for each sample
  qdf2 <- as.data.frame(qdf)#, samples=dms_no_n1_sites$sample_names) # add the metadata
  qdf2$sample <- as.vector(dms_no_n1_sites$sample_names)

  qdf3  <- merge(qdf2, dms_no_n1_sites$meta$analyses[,c("sample",site_col_name, "lat", "long")], by="sample", all.x=TRUE, all.y=FALSE)
  colnames(qdf3)[which(colnames(qdf3)==site_col_name)]<- "site"
  
  qdf4 <- qdf3
  
  qdf4$sample <- NULL
  for (i in 1:kval){
    qdf4[,i] <- as.numeric(qdf4[,i])
  }
  qdf4[,"lat"] <- as.numeric(qdf4[,"lat"])
  qdf4[,"long"] <- as.numeric(qdf4[,"long"])
  
  agg_qdf <- aggregate(. ~ site, data = qdf4, FUN = mean)
  
  scatter_map <- base_map+ labs(fill="Source\npopulation")+
  geom_scatterpie(mapping=aes(x=long, y=lat, group =site, r = diff(divxlims)/30),data =agg_qdf,
                    cols=colnames(agg_qdf)[2:(kval+1)],  alpha=1, size=0.1, colour="black", na.rm=TRUE)+
    labs(title=paste("K = ",kval))+theme(legend.position="none")
  
  # make admix plot
  qdf_long <- pivot_longer(qdf3, cols=2:(kval+1), names_to="population", values_to="Q") 

  admix_plot <- ggplot(qdf_long, 
                    aes(x=sample, #, labels=NULL),
                        y=Q, fill=population))+
    geom_bar(position="stack", stat="identity")+
    labs(y="Admixture\ncoefficient (Q)", x=element_blank(), fill="Source\npopulation")+
    facet_grid(~factor(site, levels=site_order), scales = "free_x", space = "free_x")+
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size=3),
          strip.text.x = element_text(size = 6), strip.background = element_blank(), 
          axis.title.y = element_text(size=8),
          legend.position="none")+
    scale_y_continuous(limits = c(0,1.001), expand=c(0,0))+
    theme(strip.text.x = element_text(angle = 90, size=6), panel.spacing = unit(0.07, "lines"))+
    labs(title=paste("K = ",kval))

  scatterpie_plots[[kval]] <- scatter_map
  admix_bar_plots[[kval]] <- admix_plot
}


arranged_scatterpie_plots <- scatterpie_plots[kvalrange]  # Subtract 1 because indexing is 0-based
arranged_scatterpie_plots <- wrap_plots(arranged_scatterpie_plots, ncol = 3, nrow=2)  # Change the number of columns as desired
ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/LEA_scatterpie_map.pdf"), device="pdf",
       plot = arranged_scatterpie_plots, width = 17, height = 25, units = "cm")


arranged_admix_plots <- admix_bar_plots[kvalrange]  # Subtract 1 because indexing is 0-based
arranged_admix_plots <- wrap_plots(arranged_admix_plots, ncol = 1, nrow=6) # Change the number of columns as desired
ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/LEA_barplots.pdf"), device="pdf",
       plot = arranged_admix_plots, width = 30, height = 40, units = "cm")


####################################### DIVERSITY ######################################
site_stats <- species_site_stats(dms, maf=maf_val, pop_var=species_col_name, site_var=site_col_name, missing=locus_miss)
colnames(site_stats)[1] <- site_col_name
colnames(site_stats)[14] <- species_col_name
site_stats_merged <- merge(site_stats, final_summary[,c(1:4)], by=site_col_name)

site_stats_merged <-site_stats_merged[order(match(site_stats_merged[,site_col_name], site_categories2)),]


write.xlsx(site_stats_merged, 
            paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/tables/DIVERSITY_",site_col_name,"_by_",species_col_name,".xlsx"), rowNames = FALSE)


# Find the maximum of the two size variables
het_range <- range(min(site_stats_merged[,c('obs_het', 'exp_het')], na.rm=TRUE), max(site_stats_merged[,c('obs_het', 'exp_het')], na.rm=TRUE))

ho_map <- base_map + #site_labels+
  geom_point(data = site_stats_merged, mapping = aes(x = long, y = lat, size = obs_het), alpha = 0.3, color = "red") +
  geom_point(data=site_stats_merged, mapping=aes(x=long, y=lat),size=0.1)+
  theme(legend.position = "bottom", legend.direction = "vertical", 
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(title = "HO", size="HO")+
  scale_size_continuous(limits = het_range)

he_map <- base_map + #site_labels+
  geom_point(data = site_stats_merged, mapping = aes(x = long, y = lat, size =exp_het), alpha = 0.3, color = "red") +
  geom_point(data=site_stats_merged, mapping=aes(x=long, y=lat),size=0.1)+
  theme(legend.position = "bottom", legend.direction = "vertical",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  labs(title = "HE", size="HE")+
  scale_size_continuous(limits = het_range)

fis_map <- base_map + #site_labels+
  geom_point(data=site_stats_merged, mapping=aes(x=long, y=lat, size=fis, color=fis), alpha=0.5)+ 
  scale_color_gradient2(high = "red",mid="white", midpoint = 0, low = "blue", na.value = "grey30")+
  geom_point(data=site_stats_merged, mapping=aes(x=long, y=lat),size=0.1)+
  theme(legend.position = "bottom", legend.direction = "vertical",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))+
  labs(title="FIS")

combined_stats_plot <- ( ho_map | he_map | fis_map ) / ncol(3)

ggsave(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/DIVERSITY_maps.png"), plot = combined_stats_plot, width = 30, height = 25, dpi = 600, units = "cm")

####################################### TEXT ####################################### 

allele_frequency <- calculate.AF.summary(dms)
ind_small_maf <- which((allele_frequency$P < maf_val) | allele_frequency$P > (1 - maf_val))
maf_s <- length(ind_small_maf)

dat <- data.frame("species" = as.character(species), "dataset" = as.character(dataset),
                  "d1_nloci" = ncol(d1$gt), "dms_nloci" = ncol(dms$gt),
                  "d1_nsamples" = nrow(d1$gt), "dms_nsamples" = nrow(dms$gt),
                  "maf" = maf_s)


AF.summary <- calculate.AF.summary(dms)  # Assuming dms is a data frame


sample_miss_hist <- ggplot() + geom_histogram(aes(x = rowMeans(is.na(dms$gt))), fill = "blue", color = "black") +
  labs(x = "Sample missingness", y = "Frequency")+theme(axis.title=element_text(size=8))

locus_miss_hist <- ggplot() + geom_histogram(aes(x = colMeans(is.na(dms$gt))), fill = "green", color = "black") +
  labs(x = "Locus missingness", y = "Frequency")+theme(axis.title=element_text(size=8))

af_ho_scatter <- ggplot(as.data.frame(AF.summary), aes(x = P, y = H)) + geom_point(alpha=0.1) +
  labs(x = "Allele Frequency", y = "Heterozygosity")+theme(axis.title=element_text(size=8))

QC_plots <- sample_miss_hist | locus_miss_hist | af_ho_scatter


species_line <- paste("Species: ", species, "\nDataset: ", dataset, sep = "")
pfsnps_line <- paste("Number of raw SNPs: ", length(d1$locus_names), sep = "")
snps_line <- paste("Number of quality SNPs (",locus_miss,"missing threshold): ", length(dms$locus_names), sep = "")
pfsamples_line <- paste("Number of prefilter samples: ", length(d1$sample_names), sep = "")
samples_line <- paste("Number of quality samples: ", length(dms$sample_names), sep = "")
clones_line <- paste("Number of unique genets (not filtered): ", length(unique(clones_out$genet)), sep = "")

report <- paste(species_line, pfsnps_line, snps_line,pfsamples_line,samples_line, clones_line,sep = "\n")


####################################### FINAL #######################################
                       


page1 <- ((wrap_elements(grid::textGrob(report, gp = gpar(fontsize = 16)))/QC_plots) | species_map)+plot_layout(widths = c(3,1 ))

pdf(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page1.pdf"), width = 11, height = 8.5)
plot(page1)
dev.off()


##

final_summary_df <- as.data.frame(final_summary)

if (nrow(final_summary_df) > 40) {
  # Split into chunks of 20 rows each
  chunk_size <- 40
  num_chunks <- ceiling(nrow(final_summary_df) / chunk_size)
  
  # Create a list to store tables
  table_list <- vector("list", length = num_chunks)
  
  # Loop through chunks and create tables
  for (i in 1:num_chunks) {
    start_row <- (i - 1) * chunk_size + 1
    end_row <- min(i * chunk_size, nrow(final_summary_df))
    
    table_list[[i]] <- tableGrob(final_summary_df[start_row:end_row, ], 
                                 theme = ttheme_default(base_size = 6, padding = unit(c(2, 2), "mm")))
  }
  
  # Create multiple pages with arranged tables
  pdf(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page2.pdf"))
  for (i in 1:num_chunks) {
    grid.arrange(table_list[[i]], ncol = 1)
    if (i < num_chunks) {
      cat("\f")  # Start a new page in the PDF
    }
  }
  cat("\f")
  grid.arrange(ggarrange(site_map, pca_plot_pc12_site+theme(aspect.ratio = 4/3), common.legend = TRUE, legend = legend_position)+
    plot_annotation(title = paste(species, 'site summary')))
  dev.off()
} else {
  # If less than or equal to 20 rows, plot a single table
  page2 <-(wrap_elements(gridExtra::tableGrob(as.data.frame(final_summary),theme = ttheme_default(base_size = 6, padding = unit(c(2, 2), "mm")))) |
             ggarrange(site_map, pca_plot_pc12_site+theme(aspect.ratio = 4/3), common.legend = TRUE, legend = legend_position) ) +
    plot_layout(widths = c(2,3 ))+
    plot_annotation(title = paste(species, 'site summary'))
  
  pdf(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page2.pdf"), width = 11, height = 8.5)
  plot(page2)
  dev.off()
  }


##
page3 <- combined_site_pca/combined_latitude_pca+
  plot_annotation(title = paste(species, 'PCA'))

pdf(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page3.pdf"), width = 11, height = 8.5)
plot(page3)
dev.off()

###
stats_to_print <- as.data.frame(site_stats[,c(13,14,15,1,3:4,6,12)])

if (nrow(stats_to_print) > 40) {
  # Split into chunks of 20 rows each
  chunk_size <- 40
  num_chunks <- ceiling(nrow(stats_to_print) / chunk_size)
  
  # Create a list to store tables
  table_list <- vector("list", length = num_chunks)
  
  # Loop through chunks and create tables
  for (i in 1:num_chunks) {
    start_row <- (i - 1) * chunk_size + 1
    end_row <- min(i * chunk_size, nrow(stats_to_print))
    
    table_list[[i]] <- tableGrob(stats_to_print[start_row:end_row, ], 
                                 theme = ttheme_default(base_size = 6, padding = unit(c(2, 2), "mm")))
  }
  
  # Create multiple pages with arranged tables
  pdf(paste0(species, "/outputs_", site_col_name, "_", species_col_name, "/plots/page4.pdf"))
  
  # Initial plot with annotations
  plot(combined_stats_plot +
         plot_annotation(title = paste(species, 'diversity'),
                         caption = paste0('Note: Samples are grouped by ', species_col_name, ' before filtering loci for MAF (', maf_val,') and missingness (',locus_miss,')')))
  
  # Loop for additional plots
  for (i in 1:num_chunks) {
    grid.arrange(table_list[[i]], ncol = 1)
    if (i < num_chunks) {
      cat("\f")  # Start a new page in the PDF
    }
  }
  
  dev.off()
  
} else {
  page4 <- (((wrap_elements(gridExtra::tableGrob(stats_to_print,theme = ttheme_default(base_size = 6, padding = unit(c(2, 2), "mm"))))) |
               combined_stats_plot)+
              plot_layout(widths = c(2,3)))+
    plot_annotation(title = paste(species, 'diversity'),
                    caption = paste0('Note: Samples are grouped by ', species_col_name, ' before filtering loci for MAF (', maf_val,') and missingness (',locus_miss,')'))
  
  pdf(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page4.pdf"), width = 11, height = 8.5)
  plot(page4)
  dev.off()
}
###


page5 <-  arranged_scatterpie_plots + plot_annotation( title = paste(species, 'LEA plots') )

page6 <-  (entropy_plot+theme(aspect.ratio = 2/3) | fst_manning) +
  plot_annotation( title = paste(species, 'LEA entropy plot and FST by distance'),
                   caption = paste0('Entropy note: The cross-entropy criterion assesses model fit for K populations by evaluating the prediction of masked genotypes,
                                    \n with a lower cross-entropy indicating a better model fit and aiding in the selection of the number of ancestral populations\n
                                    or the best model run for a fixed K value.'))

pdf(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/intermediate_plots.pdf"), width = 11, height = 8.5)
plot(page5)
plot(page6)
dev.off()

pdfs_to_combine <- c(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page1.pdf"),
                     paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page2.pdf"),
                     paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page3.pdf"),
                     paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page4.pdf"),
                     paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/intermediate_plots.pdf"),
                     paste0(species, "/outputs_",site_col_name,"_",species_col_name,"/plots/FST_heatmap.pdf"),
                     paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/LEA_barplots.pdf")  
)

qpdf::pdf_combine(input = pdfs_to_combine,
                  output = paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/",species,"_combined_outputs.pdf"))

file.remove(c(paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page1.pdf"),
              paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page2.pdf"),
              paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page3.pdf"),
              paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/page4.pdf"),
              paste0(species,"/outputs_",site_col_name,"_",species_col_name,"/plots/intermediate_plots.pdf")))
